plumtree-servlet/src/main/java/org/apache/tuweni/plumtree/servlet/GossipServlet.java

```diff
@@ -0,0 +1,254 @@
+import org.apache.tuweni.plumtree.PeerPruning;
+import org.apache.tuweni.plumtree.PeerRepository;
+import org.apache.tuweni.plumtree.State;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import jakarta.servlet.ServletConfig;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServlet;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GossipServlet extends HttpServlet {
+
+  private static final String PLUMTREE_SERVER_HEADER = "Plumtree-Server";
+
+  private static final Logger logger = LoggerFactory.getLogger(GossipServlet.class);
+  private static final ObjectMapper mapper = new ObjectMapper();
+
+  private static final class Message {
+
+    public MessageSender.Verb verb;
+    public String attributes;
+    public String hash;
+    public String payload;
+  }
+
+  private final int graftDelay;
+  private final int lazyQueueInterval;
+  private final MessageHashing messageHashing;
+  private final String networkInterface;
+  private final int port;
+  private final MessageListener payloadListener;
+  private final MessageValidator payloadValidator;
+  private final PeerPruning peerPruningFunction;
+  private final PeerRepository peerRepository;
+
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  private State state;
+
+  public GossipServlet(
+      int graftDelay,
+      int lazyQueueInterval,
+      MessageHashing messageHashing,
+      String networkInterface,
+      int port,
+      MessageListener payloadListener,
+      MessageValidator payloadValidator,
+      PeerPruning peerPruningFunction,
+      PeerRepository peerRepository) {
+    this.graftDelay = graftDelay;
+    this.lazyQueueInterval = lazyQueueInterval;
+    this.messageHashing = messageHashing;
+    this.networkInterface = networkInterface;
+    this.port = port;
+    this.payloadListener = payloadListener;
+    this.payloadValidator = payloadValidator == null ? (bytes, peer) -> true : payloadValidator;
+    this.peerPruningFunction = peerPruningFunction == null ? (peer) -> true : peerPruningFunction;
+    this.peerRepository = peerRepository;
+  }
+
+  private CloseableHttpClient httpclient;
+
+  @Override
+  public void init(ServletConfig config) throws ServletException {
+    super.init(config);
+    if (started.compareAndSet(false, true)) {
+      httpclient = HttpClients.createDefault();
+      state = new State(peerRepository, messageHashing, (verb, attributes, peer, hash, payload) -> {
+        Message message = new Message();
+        message.verb = verb;
+        message.attributes = attributes;
+        message.hash = hash.toHexString();
+        message.payload = payload == null ? null : payload.toHexString();
```

Comment:
Does this assume that init() is being called per request or something? How would payload not be null?

Reply:
No, this is a sender function being created as a lambda in the init function of the servlet. The sender function is used by the state to know how to send requests out.

Reply:
ahh - gotcha